shell/platform/common/cpp/text_input_model.cc

```diff
@@ -40,24 +45,100 @@ void TextInputModel::SetText(const std::string& text) {
   text_ = utf16_converter.from_bytes(text);
   selection_base_ = 0;
   selection_extent_ = 0;
+  composing_base_ = 0;
+  composing_extent_ = 0;
 }
 
 bool TextInputModel::SetSelection(size_t base, size_t extent) {
-  auto max_pos = text_.length();
-  if (base > max_pos || extent > max_pos) {
+  // Only collapsed selections are allowed while composing.
+  if (composing_ && base != extent) {
+    return false;
+  }
+  // Verify the selection is within range.
+  size_t min_pos = composing_ ? composing_start() : 0;
+  size_t max_pos = composing_ ? composing_end() : text_.length();
+  if (!InRange(base, min_pos, max_pos) || !InRange(extent, min_pos, max_pos)) {
     return false;
   }
   selection_base_ = base;
   selection_extent_ = extent;
   return true;
 }
 
-void TextInputModel::DeleteSelected() {
-  text_.erase(selection_start(), selection_end() - selection_start());
-  selection_base_ = selection_start();
+bool TextInputModel::SetComposingRange(size_t base,
+                                       size_t extent,
+                                       size_t cursor_offset) {
+  if (!composing_ || base > text_.length() || extent > text_.length()) {
+    return false;
+  }
+  composing_base_ = base;
+  composing_extent_ = extent;
+  selection_base_ = composing_start() + cursor_offset;
+  selection_extent_ = selection_base_;
+  return true;
+}
+
+void TextInputModel::BeginComposing() {
+  composing_ = true;
+  composing_base_ = selection_start();
+  composing_extent_ = composing_base_;
+}
+
+void TextInputModel::UpdateComposingText(const std::string& composing_text) {
+  std::wstring_convert<std::codecvt_utf8_utf16<char16_t>, char16_t>
+      utf16_converter;
+  std::u16string text = utf16_converter.from_bytes(composing_text);
+
+  // Preserve selection if we get a no-op update to the composing region.
+  if (text.length() == 0 && composing_base_ == composing_extent_) {
+    return;
+  }
+  if (selection_base_ != selection_extent_) {
```

Comment:
Maybe move this into DeleteSelected() so we can always just call it when it's relevant without needing a check?

Reply:
Done.